### ISSUE TITLE
Cannot use custom pragmas in `ref object`

### ISSUE THREAD

**Issue by awr1** [nim-lang/Nim#8457]
```nim
import macros
template world {.pragma.}

type
  Hello = ref object
    a: float32
    b {.world.}: int
```
output:
```
c:/nimtests $ nim c --run pragmatest2.nim
Hint: used config file 'c:\Nim\config\nim.cfg' [Conf]
Hint: system [Processing]
Hint: pragmatest2 [Processing]
Hint: macros [Processing]
pragmatest2.nim(7, 9) Error: invalid pragma: 
```

**Comment by awr1** [comment]
Any fix to this should make sure `macros.hasCustomPragma()` works with it as well.

**Comment by bluenote10** [comment]
Work-around note: The problem only applies for fields. Type level custom annotations work even for ref objects, e.g.:

```nim
type
  Hello {.world.} = ref object
    a: float32
    b: int
```

**Comment by Vindaar** [comment]
I also just stumbled on this. I noticed that defining an object as a normal object with a custom pragma and creating a ref object of that works too.

```nim
type
  HelloObj = object
    a: float32
    b {.world.} : int
  Hello = ref HelloObj
```

**Comment by krux02** [comment]
@Araq I just stumbled upon this.

**Comment by dom96** [comment]
Also just stumbled on this. Worthy of a high pri I think.

**Comment by Vindaar** [comment]
Stumbled on this again. Did some digging, but couldn't figure out how to fix it.

The problem is that here:
https://github.com/nim-lang/Nim/blob/devel/lib/core/macros.nim#L1474-L1475

the `typeInst` for a `ref object` is an `nnkSym` like "MyObj:ObjectType". And for those we cannot get the implementation via `getImpl`. It returns a `nnkNilLit`. Looking at the VM we find the `opcGetImpl` and see where that comes from:
https://github.com/nim-lang/Nim/blob/devel/compiler/vm.nim#L996
So apparently the `ast` field of such a `ref object` is nil.

Unfortunately I don't know how to fix that. :)

**Comment by alehander42** [comment]
I think this is also issue for json-serialization Lib

**Comment by ringabout** [comment]
This issue has been fixed since Nim v 1.0.2
